### ISSUE TITLE
ASSERT fiesta

### ISSUE THREAD

**Issue by lundman** [openzfsonosx/zfs#605]
With the corrections of the ASSERT code, it have come to light that we actually have a few.

Here is the list of the kernel ASSERTs seen during zfstest run. Pasted in one example stack, but they sometimes have more than one. Captured using:

```
dtrace -n 
   'assfail:entry{ printf("%s@%s:%d\n", stringof(arg0), stringof(arg1), arg2); stack(); }'
```

<details>
  <summary> assert 1 </summary>
  <p>

```
  1  71762                    assfail:entry hdr->b_l1hdr.b_freeze_cksum!=((void *)0)@arc.c:2299

              zfs`arc_buf_fill+0xcb6
              zfs`arc_buf_alloc_impl+0x79a
              zfs`arc_read+0xbbc
              zfs`dbuf_read_impl+0x29c5
              zfs`dbuf_read+0x54a
              zfs`dnode_hold_impl+0x3a0
              zfs`dnode_hold+0x34
              zfs`dmu_bonus_hold_impl+0x6a
              zfs`dmu_bonus_hold+0x34
              zfs`sa_buf_hold+0x2d
              zfs`zfs_zget_ext+0x9e
              zfs`zfs_dirent_lock+0x757
              zfs`zfs_dirlook+0x320
              zfs`zfs_lookup+0x231
              zfs`zfs_vnop_lookup+0x221
```
</details>

<details>
  <summary> assert 2 </summary>
  <p>

```
  0  71762                    assfail:entry size>0@abd.c:1223

              zfs`abd_zero_off+0x7c
              zfs`zio_write_compress+0x116c
              zfs`__zio_execute+0x5c9
              spl`taskq_thread+0x1bc
              kernel.development`call_continuation+0x17
```
</details>

<details>
  <summary> assert 3 </summary>
  <p>

```
  0  69152                    assfail:entry MUTEX_HELD(&spa_namespace_lock)@spa_misc.c:463

              zfs`spa_lookup+0x62
              zfs`spa_open_common+0x9c
              zfs`spa_open+0x31
              zfs`dsl_pool_hold+0x28
              zfs`dmu_objset_hold_flags+0x40
              zfs`dmu_objset_hold+0x29
              zfs`dsl_prop_get+0x38
              zfs`dsl_prop_get_integer+0x45
              zfs`zvol_first_open+0x1f3
              zfs`zvol_open_impl+0x69
              zfs`net_lundman_zfs_zvol_device::handleOpen(IOService*, unsigned int, void*)+0x6e
```
</details>

<details>
  <summary> assert 4 </summary>
  <p>

```
  0  69152                    assfail:entry (size_t)abd->abd_size==size@../../include/sys/abd.h:151

              zfs`abd_cmp_buf+0x3f
              zfs`abd_return_buf+0x1e6
              zfs`arc_read_done+0xbfa
              zfs`zio_done+0x32d2
              zfs`__zio_execute+0x5c9
              spl`taskq_thread+0x1bc
              kernel.development`call_continuation+0x17
```
</details>

<details>
  <summary> assert 5 </summary>
  <p>

```
  0  69152                    assfail:entry (size_t)abd->abd_size==size@../../include/sys/abd.h:144

              zfs`abd_copy_to_buf+0x3f
              zfs`abd_borrow_buf_copy+0x44
              zfs`arc_read_done+0xbc3
              zfs`zio_done+0x32d2
              zfs`__zio_execute+0x5c9
              spl`taskq_thread+0x1bc
              kernel.development`call_continuation+0x17
```
</details>

<details>
  <summary> assert 6 </summary>
  <p>

```
  1  69152                    assfail:entry hdr->b_l1hdr.b_pabd==((void *)0)@arc.c:3266

              zfs`arc_hdr_alloc_abd+0x42f
              zfs`arc_hdr_alloc+0x3e7
              zfs`arc_read+0x13c8
              zfs`dbuf_read_impl+0x29c5
              zfs`dbuf_read+0x54a
              zfs`dmu_buf_hold+0x9b
              zfs`zap_lockdir+0x5c
              zfs`zap_lookup_norm+0xab
              zfs`zap_lookup+0x6d
              zfs`zap_update+0x76
              zfs`zap_update_int_key+0x98
              zfs`dle_enqueue+0x1b8
              zfs`dsl_deadlist_insert+0x4b0
              zfs`deadlist_enqueue_cb+0x2d
              zfs`bplist_iterate+0x7c
              zfs`dsl_dataset_sync_done+0x53
              zfs`dsl_pool_sync+0x49a
              zfs`spa_sync+0x7b2
              zfs`txg_sync_thread+0x414
```
</details>

<details>
  <summary> assert 7 </summary>
  <p>

```
  0  69152                    assfail:entry spa->spa_sync_on@spa.c:6199

              zfs`spa_async_thread+0x5d
              kernel.development`call_continuation+0x17
```
</details>

<details>
  <summary> assert 8 </summary>
  <p>

```
  1  69152                    assfail:entry zp->z_is_sa@zfs_sa.c:203

              zfs`zfs_sa_get_xattr+0x22b
              zfs`zfs_vnop_getxattr+0x19c
              kernel.development`vn_getxattr+0x509
              kernel.development`vnode_get_filesec+0x38
              kernel.development`vnode_getattr+0x11e
              kernel.development`vnode_authattr+0x153
              kernel.development`chmod_vnode+0xd2
              kernel.development`chmodat+0xea
              kernel.development`chmod+0x6d
              kernel.development`unix_syscall64+0x2db

In one zfstest run there are 19503 of these. For example:

ASSERT3( writes <= zfs_vdev_async_write_max_active) failed (28318110 <= 10)
ASSERT3( writes <= zfs_vdev_async_write_max_active) failed (91560350 <= 10)
ASSERT3( writes <= zfs_vdev_async_write_max_active) failed (301470 <= 10)

```
</details>

<details>
  <summary> assert 9 </summary>
  <p>

```
  0  69152                    assfail:entry writes<=zfs_vdev_async_write_max_active@vdev_queue.c:303

              zfs`vdev_queue_max_async_writes+0x1cb
              zfs`vdev_queue_class_max_active+0x70
              zfs`vdev_queue_class_to_issue+0x115
              zfs`vdev_queue_io_to_issue+0xb7
              zfs`vdev_queue_io_done+0x94
              zfs`zio_vdev_io_done+0x172
              zfs`__zio_execute+0x5c9
              spl`taskq_thread+0x1bc
              kernel.development`call_continuation+0x17
```
</details>

<details>
  <summary> assert 10 </summary>
  <p>

```
  0  69152                    assfail:entry growtime>curtime@arc.c:4915

              zfs`arc_reclaim_thread+0x3ae
              kernel.development`call_continuation+0x17
```
</details>

This is now a **quest** for fun.

**Comment by lundman** [comment]
Reasonably sure that assert 7 is caused by assert 3 - which is a well known issue with have with IOKit calling in.

**Comment by rottegift** [comment]
Ok, we discussed assert 10, first time I've seen that go off.   As long as it's very very rare it's a harmless point of interest; if it pops up often, then we should find out who is pestering the cv_wait_timeout with a cv_signal/cv_broadcast, since an early exit from cv_wait_timeout is the only way that ASSERT3U can happen.

**Comment by lundman** [comment]
assert 1 

I'm unsure if all the arc changes recently means the asserts should be moved around, or just removed.

Seems to be known in ZOL https://github.com/zfsonlinux/zfs/issues/5721

**Comment by rottegift** [comment]
assert 1 ) the freeze code elsewhere in arc.c is guarded by checking zfs_flags & ZFS_DEBUG_MODIFY, so we do the same with https://github.com/openzfsonosx/zfs/pull/606

assert 2) rounded could == psize in zio_write_compress.  deal with that.

assert 5) fixed this as was done in vdev_file.c

**Comment by lundman** [comment]
Many / most asserts in here have been fixed in cedec9038d6d00546c5d5ce12905c0c540743fb8

There are a number more to look at, but they can get new tickets.
